enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,191 @@
+Windows service, gracefully stopping the kubelet.
+
+Once the Kubelet is stopped, the health check for the Machine associated
+with the node will be triggered after a period of time. [Machine API Operator](https://github.com/openshift/machine-api-operator/)
+will destroy and recreate the Machine, and WMCO will configure it and apply the
+new version annotation. The max amount of unavailable nodes at a time will be
+dependent on the maxUnavailable field in the [MachineSet](https://github.com/openshift/machine-api-operator/blob/master/config/machineset.yaml)
+created for Windows machines. WMCO will not render more than the specified
+amount of Windows nodes unavailable.
+
+The procedure for an upgrade is as follows:
+1) A new WMCO version is released
+2) If the current cluster version fufills the minimum Kubernetes version
+   requirement, OLM upgrades WMCO. If the cluster version is not high enough,
+   the WMCO upgrade will occur once it is.
+3) The new WMCO reconciles as usual, ensuring that all unconfigured Windows
+   Machines are configured and join the cluster as a node. Each of them are
+   given an annotation indicating the WMCO version that configured them.
+4) Each Windows node is checked for the WMCO version annotation, if the
+   annotated version of a Windows node does not match the WMCO version, the
+   kubelet on the node is stopped.
+5) When the Windows Machine is destroyed and recreated by the Machine API
+   Operator, WMCO will reconcile again and configure the VM. This will
+   repeat until all Windows nodes have been configured by the upgraded WMCO.
+
+### User Stories
+
+Stories can be found within the [Windows Upgrades epic](https://issues.redhat.com/browse/WINC-404)
+
+### Risks and Mitigations
+
+* If the cluster is upgraded, and the new version introduces a change that
+  breaks the previous version of WMCO, Windows workloads will be interrupted
+  for a period of time before WMCO is upgraded and rolls out the new nodes.
+  This will be known ahead of release and can be messaged out appropriately.
+  For unexpected complications, we will add alerts indicating nodes failing
+  after upgrades.
+* OLM does not allow for easy downgrading. The downgrade process is recommended
+  to be done by removing the operator and reinstalling the previous version. If
+  a new version of an operator causes breaking issues for a user they will have
+  to perform this manual process.
+* If a MachineHealthCheck is not set up to watch Windows Machines, the upgrade
+  process will halt until the check is created. To mitigate this, an alert will
+  be generated to inform the user when a MachineHealthCheck for the Windows
+  MachineSet needs to be created. In order to make this health check
+  discoverable by WMCO, the MachineHealthCheck must specifically select the
+  Windows MachineSet through a label selector, matching the
+  "machine.openshift.io/cluster-api-machineset" label with the name of the
```

Comment:
What would happen if there are multiple health checks with same label selector that match which one would be authoritative enough for WMCO to proceed?

Reply:
You mean a label with `machine.openshift.io/cluster-api-machineset` with `windows-machine-set` as value?

Reply:
What if there are multiple Windows machinesets? Do we want a health check for machine-set? Usually it should be per machineset.